lib/puppet/provider/package/portage.rb

```diff
@@ -127,21 +193,26 @@ def query
           result_fields.zip(match.captures) do |field, value|
             package[field] = value unless !value or value.empty?
           end
-          if package_slot
-            package[:version_available] = eix_get_version_for_slot(package[:slot_versions_available], package_slot)
-            package[:ensure] = eix_get_version_for_slot(package[:installed_slots], package_slot)
+          # dev-lang python [3.4.5] [3.5.2] [2.7.12:2.7,3.4.5:3.4] [2.7.12:2.7,3.4.5:3.4,3.5.2:3.5] https://www.python.org/ An interpreted, interactive, object-oriented programming language
+          # prefer checking versions to slots as versions are finer grained
+          if qatom[:pv]
+            package[:version_available] = eix_get_version_for_versions(package[:installable_versions_available], qatom[:pv])
+            package[:ensure] = eix_get_version_for_versions(package[:installed_versions], qatom[:pv])
+          elsif qatom[:slot]
+            package[:version_available] = eix_get_version_for_slot(package[:slot_versions_available], qatom[:slot])
+            package[:ensure] = eix_get_version_for_slot(package[:installed_slots], qatom[:slot])
```

Comment:
`ensure` and `version_available` should be reversed.  `version_available` means 'what's currently installed'.   `ensure` is what version to try and install.